### PR TITLE
Side-by-side support for Tendermint 0.34 and 0.37

### PR DESCRIPTION
**Pull request by mzabaluev**
ADR: https://github.com/informalsystems/tendermint-rs/pull/1274

A cross-chain bridge agent, such as a relayer, may need to communicate with nodes using different Tendermint versions. To enable this use with a single linked instance of `tendermint-rpc` ~/tendermint-abci~ crate, and also to disentangle tendermint-rs versioning from the versions of the Tendermint Core protocol, change `tendermint-proto` to have generated protobuf structs for all supported tendermint versions in side-by-side modules, and adapt the higher level client APIs to be able to select the version expected from the node.

* [x] Referenced an issue explaining the need for the change
* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Added entry in `.changelog/`

### CODE REVIEW

**Comment by mzabaluev** [comment]
in my understanding, an MVP of multi-version tendermint-rs may need the following changes:

- Adapt the `tendermint-abci` client to be able to talk both versions.
  - It is not used by Hermes, are there other customers for this?
  - The `Application`/ `Server` part can just serve the latest supported version; I see no use case for multi-version here.
- Add support for the RPC endpoints new in 0.37 to tendermint-rpc and conditionally disable them for the 0.34 mode, or just rely on the node to respond with an error.

**Comment by hdevalence** [comment]
I don't think it's currently possible to implement support for both 0.34 and 0.37 on `main`, because `main` does not even have support for 0.34 -- ABCI support only exists for 0.35. Before trying to add support for 0.37, the `main` branch should have support for 0.34.  Merging something like this now will generate even more massive conflicts.

**Comment by mzabaluev** [comment]
I think it should be possible to have limited legacy support for 0.34 in tendermint-rs, by generating the `v0_34` structs in `tendermint-proto`, implementing domain type conversions for these, and adapting `tendermint-rpc` to be able to interop with 0.34 nodes (which should be pretty much there due to backward compatibility). As we discussed with @thanethomson, we are not interested in multi-version adaptations for `tendermint-abci` at the moment.

**Comment by xla** [comment]
@mzabaluev What's left to be done to get this changeset completed?

**Comment by mzabaluev** [comment]
I have updated the PR.

The last significant change I wanted to try out was version-specific `Client` traits in `tendermint-rpc`, and it seems to have landed well. I think this is good to go now. Please review the changes as I'm trying to summarize them (briefly) in a changelog entry and describe them in more detail in an upcoming ADR.

**Comment by aakoshh** [comment]
Hi! I want to use the new ABCI++ features in `v0.37.0-rc2`; this PR is just perfectly timed. 

I tried to run the [abci example](https://github.com/informalsystems/tendermint-rs/tree/mikhail/multi-tc-version-support/abci) with a Tendermint I built from the source but ran into some errors immediately after starting `tendermint`.

This is the error from the ABCI process:

```console
2023-01-11T10:30:13.777266Z ERROR tendermint_abci::server: Failed to read incoming request from client 127.0.0.1:35778: error encoding protocol buffer

Caused by:
    failed to decode Protobuf message: Request.value: buffer underflow
```

Maybe it's has nothing to do with this PR, I haven't tried with the stable versions yet.

**Comment by aakoshh** [comment]
The error [above](https://github.com/informalsystems/tendermint-rs/pull/1193#issuecomment-1378558778) comes from [decode_varint](https://github.com/informalsystems/tendermint-rs/blob/mikhail/multi-tc-version-support/abci/src/codec.rs#L168-L175). According to the comments it's probably because https://github.com/tendermint/tendermint/issues/5783 has landed. 

I see that you striked out updating the `tendermint_abci` "client" to support both versions. I'm not sure if it also means not updating the server to be talked to by both versions.

File: abci/src/application.rs
Comment (aakoshh) [inline abci/src/application.rs:108]: Just to note that the defaults will probably not work (they will propose empty blocks and think [reject](https://github.com/informalsystems/tendermint-rs/pull/1193/files#diff-f98b59e5344ef9d1c053ec7d1dd516b0627d1e390a9ba8813ab004c70b340c9fR15) all proposals). The spec has [instructions](https://github.com/tendermint/tendermint/blob/v0.37.0-rc2/spec/abci/abci%2B%2B_tmint_expected_behavior.md#adapting-existing-applications-that-use-abci) for adaptation.

File: tendermint/src/v0_37/abci/request.rs
Comment (aakoshh) [inline tendermint/src/v0_37/abci/request.rs:97]: The new `PrepareProposal` and `ProcessProposal` variants are missing from `ConsensusRequest`.

File: abci/src/application.rs
Comment (mzabaluev) [reply abci/src/application.rs:108]: Right; sorry about this, `tendermint-abci` has not been my top priority for updating to 0.37. Will fix.

File: tendermint/src/v0_37/abci/request.rs
Comment (aakoshh) [reply tendermint/src/v0_37/abci/request.rs:97]: Here are the ones I found: https://github.com/informalsystems/tendermint-rs/compare/mikhail/multi-tc-version-support...aakoshh:tendermint-rs:mikhail/multi-tc-version-support

Comment (mzabaluev) [reply tendermint/src/v0_37/abci/request.rs:101]: Thank you!

File: abci/src/application.rs
Comment (mzabaluev) [reply abci/src/application.rs:108]: As I understand it, this whole default implementation only provides the most trivial default responses.
It's up to the application to fill the methods in where necessary.

**Comment by mzabaluev** [comment]
> The error [above](https://github.com/informalsystems/tendermint-rs/pull/1193#issuecomment-1378558778) comes from [decode_varint](https://github.com/informalsystems/tendermint-rs/blob/mikhail/multi-tc-version-support/abci/src/codec.rs#L168-L175). According to the comments it's possibly because [tendermint/tendermint#5783](https://github.com/tendermint/tendermint/issues/5783) has landed. I see the the new variant was already part of the codebase [once before](https://github.com/informalsystems/tendermint-rs/pull/1022/files#diff-06e2ec8e521bf5e8f9dba0ee4544a2bad5dab15d3915f10ee21fa7ef83b3f13c).
> 
> I see that you striked out updating the `tendermint_abci` "client" to support both versions. I'm not sure if it also means not updating the server to be talked to by both versions.

Would it be acceptable to only provide the latest version implementation on the server side, but optionally support the legacy wire protocol on the client?

File: abci/src/application.rs
Comment (aakoshh) [reply abci/src/application.rs:108]: Yes, I agree; what I'm trying to say is that by leaving it on defaults, filling in these methods will always be necessary. 

I can't vouch for all the other cases, but I would expect for those methods the default values represent an acceptable behaviour, like `check_tx` not finding any error, or `end_block` not changing the power table. 

But with these two new methods, the defaults are not accepted by Tendermint:
* If we return default from `prepare_proposal`, that effectively will remove all transactions Tendermint wanted to propose, resulting in an empty block instead. 
* If then we return `ProcessProposal::Unknown` (which is the default, although it could be `Accept`) instead of `ProcessProposal::Accept`, we get the following error: `CONSENSUS FAILURE!!!                         module=consensus err="ProcessProposal responded with status UNKNOWN"`. That is because Tendermint now considers us insane: we just asked it to propose a block we ourselves do not accept.

So, if we want to have a transition to v0.37 where users of the `Application` trait don't have to override unless they have something to change from the default behaviour, and they expect Tendermint to just work, then the default method implementations have a bit more work to do than usual.

(NB the specs explicitly say that TC _may_ pass more transactions in `PrepareProposal` than what fits into the maximum size; otherwise the implementation would be a trivial copy).

**Comment by aakoshh** [comment]
> Would it be acceptable to only provide the latest version implementation on the server side, but optionally support the legacy wire protocol on the client?

This must be a question to your fellow maintainers, but here's my 2 cents:

As I understand the client is mostly for testing, because it directly talks to the ABCI interface, while normally only TC does that. So I would assume it only has to support what the server does, there's no need for optional support for alternative formats.

If the server only supports the new format, then developers who don't want to or can't yet move to TC 0.34 _and_ they use the ABCI server have to refrain from updating to the `tendermint-rs`, or they would not be able to run their blockchain any more. 

From what I gathered, the multi-version support is primarily meant for relayers who need to talk to different versions of TC at the same time, but they would call the RPC interface rather than the ABCI directly, so for them it doesn't matter.

I don't know what's the most tasteful solution (a feature flag?), but a similar question will arise in `tower-abci` as well.

**Comment by mzabaluev** [comment]
I have changed the encoding to unsigned varint in https://github.com/informalsystems/tendermint-rs/pull/1193/commits/620aa99114d2495d827b292cb43ba69c07613857
This corresponds to the earlier update to the 0.37 protobuf, without alternatives, in `tendermint-abci`.
Please comment if alternative approaches are needed; I will also add this to the upcoming ADR describing these changes.

**Comment by romac** [comment]
It would be great if we could add some integration tests against 0.34 and 0.37, for the `tendermint-rpc` crate on top of the existing fixture-based tests we have already in `tendermint-rpc/tests`.

What do you think @mzabaluev @thanethomson?

**Comment by mzabaluev** [comment]
We should, at minimum, have fixtures for the new requests in 0.37, and test that `v0_37::Client` interoperates with these.
Old fixtures can be tested against both `Client` trait implementations, but the extra code that will exercise is pretty trivial.

**Comment by mzabaluev** [comment]
@thanethomson What would it take to re-acquire fixtures from Gaia or simapp based on tendermint 0.37 with rpc-probe, as was done in e562d4d7e74710f9bd6f80330a5f1ce9bb9d2ed6?

**Comment by thanethomson** [comment]
> What would it take to re-acquire fixtures from Gaia or simapp based on tendermint 0.37 with rpc-probe

You'd need a version of Gaia integrated with the latest v0.47.x release of the Cosmos SDK (https://github.com/cosmos/cosmos-sdk/releases/tag/v0.47.0-rc1). I'd recommend speaking with @mpoke about this.

**Comment by ljoss17** [comment]
Blocking https://github.com/informalsystems/hermes/issues/3002

File: rpc/src/endpoint/tx_search.rs
Comment (mzabaluev) [inline rpc/src/endpoint/tx_search.rs:52]: This is annoying to adapt to in client code, while the data type the client has to deal with does not really have to be different: the only problem we're solving with the generics is differences in serialization. I'm inclined to get this back to `Response` using domain types, with `DialectResponse<Ev>` alongside for JSON deserialization. This, however, will require changes to the `Request` trait.

**Comment by romac** [comment]
Left some initial comments after a quick skim for us to discuss during the sync review. Will do another, deeper pass over the PR before the sync review.

File: abci/src/codec.rs
Comment (hdevalence) [reply abci/src/codec.rs:169]: No, it's a breaking change from 0.34. It was originally supposed to be included in 0.35. I don't know if it's slated for inclusion in 0.37.

File: proto/src/serializers.rs
Comment (mzabaluev) [reply proto/src/serializers.rs:58]: No detailed memory of this now, but these serializers should only used internally by the serde implementations derived in the crate. The contents of the module have changed in breaking ways, too, so I think it's better to hide it.

File: abci/src/codec.rs
Comment (mzabaluev) [reply abci/src/codec.rs:169]: The encoding has changed in 0.37, and in this PR, `tendermint-abci` is converted completely to 0.37 with no compatibility support. We can revert all of this and declare `tendermint-abci` at end of life if that's better.

Comment (mzabaluev) [reply abci/src/codec.rs:169]: We could also choose to commit to multi-version support here as well, but this will require more work.

File: abci/src/application.rs
Comment (romac) [reply abci/src/application.rs:108]: @aakoshh Would that work for you or would you rather leave out the default impls?

File: proto/src/serializers.rs
Comment (romac) [reply proto/src/serializers.rs:58]: For the sake of consistency, let's perhaps make them all public here but make the `serializers` module private.

File: abci/src/application.rs
Comment (aakoshh) [reply abci/src/application.rs:108]: Thanks for asking @romac, my vote goes for having a default implementation that copies transactions and accepts proposals, unless overridden, like @mzabaluev said. This is so people don't have to come up with boilerplate to do this every time they implement the trait, similar to how they don't have to override `check_tx` unless they want to actually perform some logic. 

For example I forked `tower-abci` to point it at this branch of `tendermint-rs` and had to implement this logic [for the kvstore example](https://github.com/consensus-shipyard/tower-abci/pull/1/files#diff-3064912e722a6c3a27e57368a923034ba4dff6408ddfc4923ef4e00ed68b38bc), where it seems a bit accidental among all the other `Default::default()`s. After that I took inspiration from the `Application` trait in `tendermint-rs` and provided an [async version](https://github.com/consensus-shipyard/fendermint/blob/master/fendermint/abci/src/application.rs) which takes care of these defaults, so I think it's the sensible thing to do.

Comment (mzabaluev) [reply abci/src/application.rs]: Thank you @aakoshh for your feedback.
The more elaborate (but still very naive) default behaviours have now been implemented in https://github.com/informalsystems/tendermint-rs/pull/1193/commits/7e79039558b50fcdd3ca47512b4e58ed6624132f

File: rpc/src/client/transport/http.rs
Comment (mzabaluev) [inline rpc/src/client/transport/http.rs:59]: This is not the best design for maintainability, and perhaps a better solution would be a builder pattern for HttpClient.

File: rpc/src/client/transport/websocket.rs
Comment (mzabaluev) [inline rpc/src/client/transport/websocket.rs:65]: Also a builder would be better here, instead of introducing another struct with public members.

File: tools/proto-compiler/src/constants.rs
Comment (mzabaluev) [inline tools/proto-compiler/src/constants.rs:26]: This is the last tag in Tendermint Core where the proto files had bundled Google well known types, before switching to buf to refer to dependencies. Adapting to buf will require more changes to the proto-compiler tool, and I have already done this already once in an older branch.

Here I'm leaving the migration to buf, and switching to CometBFT, as follow-on tasks.
There is no difference in protobuf message formats with the latest state in CometBFT branch `v0.37.x` (as per commit d26bdfe1d424ff24c9676fee5bef9a180870315e).

File: rpc/src/client/compat.rs
Comment (romac) [inline rpc/src/client/compat.rs:12]: What's the reasoning behind using `Latest` here instead of `V0_37`? 

Having the datatype variant change meaning over releases sounds brittle to me, and it would be more explicit to have it denote the actual release that is supported.

If there are no good reasons against it, how about we rename this to `V0_37` and add a `CompatMode::latest()` method that returns the latest compat mode. Let's also change the `Default` instance to return `CompatMode::latest()`.

Comment (mzabaluev) [reply rpc/src/client/compat.rs:12]: Good idea. There may be a case for the `Latest` variant alongside both explicitly selective versions, but perhaps this will produce complications on matching. So let's do "latest" as an associated const or fn.

Comment (mzabaluev) [reply rpc/src/client/compat.rs]: Redone in https://github.com/informalsystems/tendermint-rs/pull/1193/commits/f3e161e02b4bc42a9c60a369a4b0b7ecd779da99

File: rpc/src/client/transport/http.rs
Comment (mzabaluev) [reply rpc/src/client/transport/http.rs]: Redone in https://github.com/informalsystems/tendermint-rs/pull/1193/commits/84d9cad8c8e41e9e0343633a06ae8288101b0ad3

File: rpc/src/client/transport/websocket.rs
Comment (mzabaluev) [reply rpc/src/client/transport/websocket.rs]: Redone in https://github.com/informalsystems/tendermint-rs/pull/1193/commits/84d9cad8c8e41e9e0343633a06ae8288101b0ad3

File: rpc/src/client/bin/main.rs
Comment (romac) [inline rpc/src/client/bin/main.rs:495]: Shall we discover the compat and then use the appropriate dialect here?
I don't who is or has ever used this rpc binary, but if someone is they may be surprised if they try running it against a 0.34 node. Or maybe it'll just work since one can decode a 0.37 event using the 0.34 schema (but not the other way around).

File: rpc/src/client/transport/http.rs
Comment (romac) [inline rpc/src/client/transport/http.rs:54]: Very nice!

**Review by romac — APPROVED** [review]
Amazing work there @mzabaluev! I think we can finally get this in :)

I have one question left, but I don't believe it should block this PR from being merged.

File: rpc/src/client/transport/http.rs
Comment (mzabaluev) [inline rpc/src/client/transport/http.rs:88]: Hm, OK. I did not think it worthwhile to change to the less usable self-consuming builder, just to save a few clone operations on small data structures.
Or is there a bigger reason?

File: rpc/src/client/bin/main.rs
Comment (mzabaluev) [reply rpc/src/client/bin/main.rs:495]: This is only used to display the events by converting them back into JSON. I think it's actually better to show events in plain text even if the original JSON is from 0.34.

That said, this binary does not do version discovery at all now, so the clients would always decode as 0.37. This we shall change.

Comment (mzabaluev) [reply rpc/src/client/bin/main.rs:529]: Added version discovery for the CLI in https://github.com/informalsystems/tendermint-rs/pull/1193/commits/2d4a3336458f3d0cea6ae90928ab3f2aabdfb4ef

File: rpc/src/client/transport/http.rs
Comment (romac) [reply rpc/src/client/transport/http.rs]: The `&mut pattern` does work very well for our use in the event monitor, and since that's the main consumer of the library I figured we could switch to a self-consuming one. It's slightly less flexible in the sense that one may have to re-assign it to variable when setting things conditionally but otherwise it makes for a more fluent API in our current use case. The lack of cloning is just a side benefit from consuming the builder at the end, as most builders do.

**Review by romac — APPROVED** [review]

**Pull request merged**
